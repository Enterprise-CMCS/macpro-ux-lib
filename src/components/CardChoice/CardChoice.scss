@use "../../assets/theme/macbis-colors" as c;

.card-choice {
  cursor: pointer;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  padding: 26px;
  text-decoration: none;
  &.card-choice--bordered {
    border: solid 2px c.$gray-lighter;
    border-top: none;
    padding: 24px;
  }
  &.card-choice--dark {
    background-color: c.$gray-lightest;
  }
  &:hover {
    background-color: c.$primary-alt-lightest;
    border: solid 2px c.$primary-alt-dark;
    padding: 24px;
    &.card-choice--bordered {
      margin-top: -2px;
    }
  }
  &:focus {
    outline: 0.25rem solid #2491ff;
    outline-offset: 0;
    position: relative;
  }
  .content {
    line-height: 24px;
    .heading {
      color: c.$primary;
      font-weight: 700;
      font-size: 18px;
    }
    .body {
      color: c.$gray;
      font-weight: 400;
      font-size: 16px;
    }
  }
  .select {
    display: flex;
    flex-direction: column;
    justify-content: center;
    color: c.$primary;
    span {
      display: flex;
      flex-direction: row;
      align-items: center;
      svg {
        color: c.$primary;
      }
    }
  }
}
