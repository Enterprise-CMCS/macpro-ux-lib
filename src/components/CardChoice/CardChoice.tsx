import { Icon } from "components/Icon/Icon";
import React, { PropsWithChildren } from "react";

type IntrinsicElements = JSX.IntrinsicElements["a"];

interface Props extends IntrinsicElements {
  actionText?: string;
  bodyText?: string;
  bordered?: boolean;
  className?: string;
  darkBG?: boolean;
  headingText?: string;
}

/**
 * CardChoice Component
 *
 * CardChoice is wrapped by an anchor tag. All unspecified props will be provided to the anchor.
 * @param {string}             actionText     Optional text prompt to be displayed left of the navigation arrow.
 * @param {boolean}            bordered       Renders a border around the CardChoice. Applying borderd on CardChoice will override the CardChoiceGroup.
 * @param {string}             bodyText       Text to be rendered in the body.
 * @param {React.ReactNode}    children       Children provided will be rendered below the bodyText.
 * @param {string}             className      Additional classes that can be applied to the root element.
 * @param {boolean}            darkBG         Renders the CardChoice with a darker background. Applying darkBG on CardChoice will override the CardChoiceGroup. CardChoiceGroup has an option for alternatingBG.
 * @param {string}             headingText    Bolded heading text displayed at the top of the CardChoice.
 * @param {string}             href           href provided to the anchor.
 * @param                      onClick        onClick provided to the anchor.
 */
export const CardChoice: React.FC<PropsWithChildren<Props>> = ({
  actionText,
  bodyText,
  bordered,
  children,
  className,
  darkBG,
  headingText,
  href,
  onClick,
  ...rest
}) => {
  const classes = `card-choice${darkBG ? " card-choice--dark" : ""}${
    bordered ? " card-choice--bordered" : ""
  }${className ? ` ${className}` : ""}`;
  return (
    <a {...rest} className={classes} href={href} onClick={onClick}>
      <div className="content">
        <p className="heading">{headingText}</p>
        <span className="body">
          <p>{bodyText}</p>
          {children}
        </span>
      </div>
      <div className="select">
        <span>
          {actionText}
          <Icon name="navigate_next" />
        </span>
      </div>
    </a>
  );
};
