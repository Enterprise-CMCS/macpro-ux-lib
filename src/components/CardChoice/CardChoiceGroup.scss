@use "../../assets/theme/macbis-colors" as c;
.card-choice-group {
  .gradient-cap {
    background: linear-gradient(to right, c.$primary, c.$primary-alt);
    border-radius: 4px 4px 0 0;
    display: inline-block;
    height: 8px;
    width: 100%;
    margin: 0;
    position: relative;
    top: 4px;
  }
}
