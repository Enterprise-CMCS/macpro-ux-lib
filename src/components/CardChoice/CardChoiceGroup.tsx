import React, { Children, PropsWithChildren } from "react";

type IntrinsicElements = JSX.IntrinsicElements["div"];

interface Props extends IntrinsicElements {
  alternatingBG?: boolean;
  bordered?: boolean;
}

/**
 * CardChoiceGroup Component
 * @param {boolean}            [alternatingBG]    CardChoice children are displayed with an alternating background color. The darkBG property on an individual CardChoice will override this property.
 * @param {boolean}            [bordered]         All CardChoice children are displayed with a gray border. The bordered property on an individual CardChoice will override this property.
 * @param {React.ReactNode}    [children]         CardChoice children to be rendered.
 */
export const CardChoiceGroup: React.FC<PropsWithChildren<Props>> = ({
  alternatingBG = false,
  bordered = false,
  children,
  className,
  ...rest
}) => {
  const arrayChildren = Children.toArray(children);
  return (
    <div
      className={`card-choice-group${className ? ` ${className}` : ""}`}
      {...rest}
    >
      <span className="gradient-cap"></span>
      {Children.map(arrayChildren, (child, idx) => {
        if (React.isValidElement(child)) {
          return React.cloneElement(child as React.ReactElement, {
            bordered: child.props.bordered ?? bordered,
            darkBG: child.props.darkBG ?? (alternatingBG && idx % 2),
          });
        }
      })}
    </div>
  );
};
