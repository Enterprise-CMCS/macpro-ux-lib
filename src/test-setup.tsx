import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
