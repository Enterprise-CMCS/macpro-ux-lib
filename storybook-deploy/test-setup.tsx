import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
