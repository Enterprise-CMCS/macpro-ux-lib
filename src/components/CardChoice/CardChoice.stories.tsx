import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { CardChoice } from "./CardChoice";

export default {
  title: "COMPONENTS/CardChoice/CardChoice",
  component: CardChoice,
  args: {
    actionText: "Select",
    bodyText: "Nothing special. Just a standard UI component.",
    headingText: "This is a CardChoice Component",
    href: "",
  },
  argTypes: {
    actionText: {
      description:
        "Optional text prompt to be displayed left of the navigation arrow.",
    },
    bordered: {
      description:
        "Renders a border around the CardChoice. Applying borderd on CardChoice will override the CardChoiceGroup.",
    },
    bodyText: { description: "Text to be rendered in the body." },
    children: {
      description:
        "Children provided will be rendered below the bodyText.\n\n`React.ReactNode`",
    },
    className: {
      description:
        "Additional classes that can be applied to the root element.",
    },
    darkBG: {
      description:
        "Renders the CardChoice with a darker background. Applying darkBG on CardChoice will override the CardChoiceGroup. CardChoiceGroup has an option for alternatingBG.",
    },
    headingText: {
      description:
        "Bolded heading text displayed at the top of the CardChoice.",
    },
    href: { description: "href provided to the Link.", type: "string" },
    onClick: { description: "onClick provided to the Link.", type: "function" },
  },
} as ComponentMeta<typeof CardChoice>;

const Template: ComponentStory<typeof CardChoice> = ({ ...rest }) => (
  <CardChoice {...rest} />
);

export const Default = Template.bind({});
Default.args = {};

export const Bordered = Template.bind({});
Bordered.args = { bordered: true };

export const DarkBackground = Template.bind({});
DarkBackground.args = { darkBG: true };
