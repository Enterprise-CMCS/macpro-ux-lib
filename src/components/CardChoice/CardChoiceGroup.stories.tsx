import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { CardChoice } from "./CardChoice";
import { CardChoiceGroup } from "./CardChoiceGroup";

export default {
  title: "COMPONENTS/CardChoice/CardChoiceGroup",
  component: CardChoice,
  argTypes: {},
  args: {},
} as ComponentMeta<typeof CardChoiceGroup>;

const Template: ComponentStory<typeof CardChoiceGroup> = ({ ...rest }) => (
  <CardChoiceGroup>
    <CardChoice bordered />
    <CardChoice bordered />
    <CardChoice />
    <CardChoice darkBG />
    <CardChoice />
  </CardChoiceGroup>
);

export const Default = Template.bind({});
Default.args = {};
