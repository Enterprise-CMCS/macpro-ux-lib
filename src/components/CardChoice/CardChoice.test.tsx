import React from "react";
import { CardChoice } from "./CardChoice";
import { CardChoiceGroup } from "./CardChoiceGroup";
import { fireEvent, screen, render } from "../../test-setup";

describe("Tests for the CardChoice component", () => {
  const headingText = "A CardChoice Component";
  const bodyText =
    "This is a standard CardChoice component. It is both an informational and navigational tool.";
  it("should render with defaults", () => {
    const { container } = render(
      <CardChoice headingText={headingText} bodyText={bodyText} />
    );
    const component = container.firstElementChild!;

    expect(component.className).toBe("card-choice");
    expect(component.getAttribute("href")).toBe(null);
    expect(component.getAttribute("aria-label")).toBe(
      "Navigation Card for: A CardChoice Component"
    );
    expect(screen.getByText(headingText)).toBeVisible();
    expect(screen.getByText(bodyText)).toBeVisible();

    // check no actionText visible
    const selectSpan =
      component.getElementsByClassName("select")[0].firstElementChild;
    expect(selectSpan?.firstElementChild?.tagName).toBe("svg");
  });

  it("should render body text and children", () => {
    const { container } = render(
      <CardChoice headingText={headingText} bodyText={bodyText}>
        <ul className="test-card-choice-child">
          <li>Item 1</li>
          <li>Item 2</li>
        </ul>
      </CardChoice>
    );

    // body should have bodyText and children; bodyText should appear first
    const body = container.getElementsByClassName("body")[0];
    expect(body.children.length).toBe(2);
    expect(body.children[0].innerHTML).toBe(bodyText);
    expect(body.children[1].className).toBe("test-card-choice-child");
  });

  it("should have alternate dark background", () => {
    const { container } = render(<CardChoice darkBG />);
    const component = container.firstElementChild!;
    expect(component.className).toBe("card-choice card-choice--dark");
  });

  it("should have border", () => {
    const { container } = render(<CardChoice darkBG bordered />);
    const component = container.firstElementChild!;
    expect(component.className).toBe(
      "card-choice card-choice--dark card-choice--bordered"
    );
  });

  it("should have a custom className", () => {
    const { container } = render(
      <CardChoice darkBG bordered className="test-class" />
    );
    const component = container.firstElementChild!;
    expect(component.className).toBe(
      "card-choice card-choice--dark card-choice--bordered test-class"
    );
  });

  it("should have actionText", () => {
    const { container } = render(<CardChoice actionText="Click Me" />);
    const selectSpan =
      container.getElementsByClassName("select")[0].firstElementChild;

    // should have text and svg image
    expect(selectSpan?.textContent).toBe("Click Me");
    expect(selectSpan?.firstElementChild?.tagName).toBe("svg");
  });

  it("should have onClick event", () => {
    const mockOnClick = jest.fn();
    const { container } = render(<CardChoice onClick={mockOnClick} />);

    fireEvent.click(container.firstElementChild!);
    expect(mockOnClick).toHaveBeenCalled();
  });

  it("should have custom aria-label", () => {
    const { container } = render(
      <CardChoice ariaLabel="my aria-label is great" />
    );
    expect(container.firstElementChild!.getAttribute("aria-label")).toBe(
      "my aria-label is great"
    );
  });

  it("should have basic aria-label when no details provided", () => {
    const { container } = render(<CardChoice />);
    expect(container.firstElementChild!.getAttribute("aria-label")).toBe(
      "Navigation Card"
    );
  });
});

describe("Tests for the CardChoiceGroup component", () => {
  it("should render with defaults", () => {
    const { container } = render(<CardChoiceGroup />);
    const component = container.firstElementChild!;
    expect(component.className).toBe("card-choice-group");
    expect(component.children.length).toBe(1);
    expect(component.firstElementChild?.tagName).toBe("SPAN");
    expect(component.firstElementChild?.className).toBe("gradient-cap");
  });

  it("should render with custom className", () => {
    const { container } = render(
      <CardChoiceGroup className="test-ccg-class" />
    );
    expect(container.firstElementChild?.className).toBe(
      "card-choice-group test-ccg-class"
    );
  });
});

describe("Tests for the CardChoice and CardChoiceGroup components", () => {
  it("should render", () => {
    const { container } = render(
      <CardChoiceGroup>
        <CardChoice />
        <CardChoice />
        <CardChoice />
      </CardChoiceGroup>
    );

    // 4 children - span and 3 CardChoice
    expect(container.firstElementChild!.children.length).toBe(4);
  });

  it("CardChoice elements should have alternating BG color", () => {
    const { container } = render(
      <CardChoiceGroup alternatingBG>
        <CardChoice />
        <CardChoice />
        <CardChoice />
      </CardChoiceGroup>
    );
    const cardChoices = [...container.getElementsByClassName("card-choice")];
    expect(cardChoices.length).toBe(3);

    expect(cardChoices[0].className).toBe("card-choice");
    expect(cardChoices[1].className).toBe("card-choice card-choice--dark");
    expect(cardChoices[2].className).toBe("card-choice");
  });

  it("CardChoice darkBG prop should override CardChoiceGroup alternatingBG", () => {
    const { container } = render(
      <CardChoiceGroup alternatingBG>
        <CardChoice darkBG />
        <CardChoice darkBG={false} />
        <CardChoice />
      </CardChoiceGroup>
    );
    const cardChoices = [...container.getElementsByClassName("card-choice")];
    expect(cardChoices.length).toBe(3);

    expect(cardChoices[0].className).toBe("card-choice card-choice--dark");
    expect(cardChoices[1].className).toBe("card-choice");
    expect(cardChoices[2].className).toBe("card-choice");
  });

  it("all CardChoice elements should have a border", () => {
    const { container } = render(
      <CardChoiceGroup bordered>
        <CardChoice />
        <CardChoice />
        <CardChoice />
      </CardChoiceGroup>
    );
    const cardChoices = [...container.getElementsByClassName("card-choice")];
    expect(cardChoices.length).toBe(3);

    expect(cardChoices[0].className).toBe("card-choice card-choice--bordered");
    expect(cardChoices[1].className).toBe("card-choice card-choice--bordered");
    expect(cardChoices[2].className).toBe("card-choice card-choice--bordered");
  });

  it("CardChoice bordered prop should override CardChoiceGroup bordered", () => {
    const { container } = render(
      <CardChoiceGroup bordered>
        <CardChoice />
        <CardChoice />
        <CardChoice bordered={false} />
      </CardChoiceGroup>
    );
    const cardChoices = [...container.getElementsByClassName("card-choice")];
    expect(cardChoices.length).toBe(3);

    expect(cardChoices[0].className).toBe("card-choice card-choice--bordered");
    expect(cardChoices[1].className).toBe("card-choice card-choice--bordered");
    expect(cardChoices[2].className).toBe("card-choice");
  });

  it("should match snapshot", () => {
    const { container } = render(
      <CardChoiceGroup alternatingBG bordered>
        <CardChoice headingText="CardChoice" bodyText="This is the body." />
        <CardChoice headingText="CardChoice" bodyText="This is the body." />
        <CardChoice headingText="CardChoice" bodyText="This is the body." />
      </CardChoiceGroup>
    );
    expect(container).toMatchSnapshot();
  });
});
