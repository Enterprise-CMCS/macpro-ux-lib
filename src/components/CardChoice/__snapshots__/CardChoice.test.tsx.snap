// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the CardChoice and CardChoiceGroup components should match snapshot 1`] = `
<div>
  <div
    class="card-choice-group"
  >
    <span
      class="gradient-cap"
    />
    <a
      aria-label="Navigation Card for: CardChoice"
      class="card-choice card-choice--bordered"
    >
      <div
        class="content"
      >
        <h6
          aria-level="6"
          class="heading"
          role="heading"
        >
          CardChoice
        </h6>
        <span
          class="body"
        >
          <p>
            This is the body.
          </p>
        </span>
      </div>
      <div
        class="select"
      >
        <span
          aria-hidden="true"
        >
          <svg
            aria-hidden="true"
            aria-label="navigate next icon"
            class="usa-icon--size-3"
            color="black"
            fill="currentColor"
            role="img"
          >
            <use
              href="sprite.svg#navigate_next"
            />
          </svg>
        </span>
      </div>
    </a>
    <a
      aria-label="Navigation Card for: CardChoice"
      class="card-choice card-choice--dark card-choice--bordered"
    >
      <div
        class="content"
      >
        <h6
          aria-level="6"
          class="heading"
          role="heading"
        >
          CardChoice
        </h6>
        <span
          class="body"
        >
          <p>
            This is the body.
          </p>
        </span>
      </div>
      <div
        class="select"
      >
        <span
          aria-hidden="true"
        >
          <svg
            aria-hidden="true"
            aria-label="navigate next icon"
            class="usa-icon--size-3"
            color="black"
            fill="currentColor"
            role="img"
          >
            <use
              href="sprite.svg#navigate_next"
            />
          </svg>
        </span>
      </div>
    </a>
    <a
      aria-label="Navigation Card for: CardChoice"
      class="card-choice card-choice--bordered"
    >
      <div
        class="content"
      >
        <h6
          aria-level="6"
          class="heading"
          role="heading"
        >
          CardChoice
        </h6>
        <span
          class="body"
        >
          <p>
            This is the body.
          </p>
        </span>
      </div>
      <div
        class="select"
      >
        <span
          aria-hidden="true"
        >
          <svg
            aria-hidden="true"
            aria-label="navigate next icon"
            class="usa-icon--size-3"
            color="black"
            fill="currentColor"
            role="img"
          >
            <use
              href="sprite.svg#navigate_next"
            />
          </svg>
        </span>
      </div>
    </a>
  </div>
</div>
`;
